File: frontend/src/game/engine/GameEngine.ts
import type { BoardState, Move } from './types';
import { createEmptyBoard, cloneBoard, BLACK } from './types';
import type { GameRules } from './rules/GameRules';
import { StandardGomokuRules } from './rules/StandardGomokuRules';

export class GameEngine {
  private _board: BoardState;
  private _currentPlayer: number;
  private _moves: Move[];
  private _winner: number | null;
  private _rules: GameRules;

  constructor(rules?: GameRules) {
    this._rules = rules || new StandardGomokuRules();
    this._board = createEmptyBoard();
    this._currentPlayer = BLACK;
    this._moves = [];
    this._winner = null;
  }

  get board(): BoardState {
    return this._board;
  }

  get currentPlayer(): number {
    return this._currentPlayer;
  }

  get moves(): Move[] {
    return [...this._moves];
  }

  get winner(): number | null {
    return this._winner;
  }

  get rules(): GameRules {
    return this._rules;
  }

  get isGameOver(): boolean {
    return this._winner !== null;
  }

  get moveCount(): number {
    return this._moves.length;
  }

  reset(): void {
    this._board = createEmptyBoard();
    this._currentPlayer = BLACK;
    this._moves = [];
    this._winner = null;
  }

  placeStone(row: number, col: number): boolean {
    if (this._winner !== null) return false;
    if (!this._rules.isValidMove(this._board, row, col, this._currentPlayer)) return false;

    this._board[row][col] = this._currentPlayer;
    const move: Move = { row, col, player: this._currentPlayer, timestamp: Date.now() };
    this._moves.push(move);

    if (this._rules.checkWin(this._board, row, col, this._currentPlayer)) {
      this._winner = this._currentPlayer;
      return true;
    }

    if (this._rules.isBoardFull(this._board)) {
      this._winner = 0;
      return true;
    }

    this._currentPlayer = this._currentPlayer === 1 ? 2 : 1;
    return true;
  }

  checkWinAt(row: number, col: number): boolean {
    if (this._board[row][col] === 0) return false;
    return this._rules.checkWin(this._board, row, col, this._board[row][col]);
  }

  setRules(rules: GameRules): void {
    this._rules = rules;
  }

  clone(): GameEngine {
    const engine = new GameEngine(this._rules);
    engine._board = cloneBoard(this._board);
    engine._currentPlayer = this._currentPlayer;
    engine._moves = [...this._moves];
    engine._winner = this._winner;
    return engine;
  }
}

File: backend/src/main/java/com/gobang/engine/GameEngine.java
package com.gobang.engine;

import com.gobang.engine.rules.GameRules;
import com.gobang.engine.rules.StandardGomokuRules;
import com.gobang.model.GameState;

import java.util.ArrayList;

public class GameEngine {
    private GameState state;
    private GameRules rules;

    public GameEngine(String gameId) {
        this(gameId, new StandardGomokuRules());
    }

    public GameEngine(String gameId, GameRules rules) {
        this.rules = rules;
        this.state = new GameState();
        this.state.setId(gameId);
        this.state.setBoard(GameConstants.createEmptyBoard());
        this.state.setCurrentPlayer(GameConstants.BLACK);
        this.state.setMoves(new ArrayList<>());
        this.state.setWinner(null);
        this.state.setCreatedAt(java.time.LocalDateTime.now().toString());
    }

    public GameState getState() {
        return state;
    }

    public GameRules getRules() {
        return rules;
    }

    public void setRules(GameRules rules) {
        this.rules = rules;
    }

    public int[][] getBoard() {
        return state.getBoard();
    }

    public int getCurrentPlayer() {
        return state.getCurrentPlayer();
    }

    public Integer getWinner() {
        return state.getWinner();
    }

    public boolean isGameOver() {
        return state.getWinner() != null;
    }

    public boolean placeStone(int row, int col) {
        if (isGameOver()) return false;
        if (!rules.isValidMove(state.getBoard(), row, col, state.getCurrentPlayer())) return false;

        state.getBoard()[row][col] = state.getCurrentPlayer();
        state.getMoves().add(new GameState.Move(row, col, state.getCurrentPlayer(), System.currentTimeMillis()));

        if (rules.checkWin(state.getBoard(), row, col, state.getCurrentPlayer())) {
            state.setWinner(state.getCurrentPlayer());
            return true;
        }

        if (rules.isBoardFull(state.getBoard())) {
            state.setWinner(0);
            return true;
        }

        state.setCurrentPlayer(GameConstants.opponentOf(state.getCurrentPlayer()));
        return true;
    }

    public boolean checkWinAt(int row, int col) {
        if (state.getBoard()[row][col] == GameConstants.EMPTY) return false;
        return rules.checkWin(state.getBoard(), row, col, state.getBoard()[row][col]);
    }

    public void reset() {
        state.setBoard(GameConstants.createEmptyBoard());
        state.setCurrentPlayer(GameConstants.BLACK);
        state.setMoves(new ArrayList<>());
        state.setWinner(null);
    }

    public GameEngine clone() {
        GameEngine engine = new GameEngine(state.getId(), rules);
        engine.state.setBoard(GameConstants.cloneBoard(state.getBoard()));
        engine.state.setCurrentPlayer(state.getCurrentPlayer());
        engine.state.setMoves(new ArrayList<>(state.getMoves()));
        engine.state.setWinner(state.getWinner());
        engine.state.setCreatedAt(state.getCreatedAt());
        return engine;
    }
}
